compositor_pipeline/src/audio_mixer/internal_audio_mixer.rs

```diff
@@ -109,45 +126,159 @@ impl InternalAudioMixer {
+
+            // check if we need to drop samples at the beginning
+            let mut start_range = 0;
+            if expected_next_sample_start_pts > input_samples.start_pts + max_error {
+                let time_to_remove_from_start =
+                    expected_next_sample_start_pts.saturating_sub(input_samples.start_pts);
+                let samples_to_remove_from_start = time_to_sample_count(time_to_remove_from_start);
+                if batch_index != 0 {
+                    // We should only drop samples in the first batch.
+                    warn!(
+                        "Decoder produced overlapping samples. Dropping {} samples.",
+                        samples_to_remove_from_start
+                    );
+                }
+                start_range = samples_to_remove_from_start;
+            };
+
+            // check if we need to drop samples at the end
+            let mut end_range = input_samples.len();
+            if input_samples.end_pts > end_pts + max_error {
+                let desired_duration = end_pts.saturating_sub(expected_next_sample_start_pts);
+                let desired_sample_count = time_to_sample_count(desired_duration);
+                end_range = start_range + desired_sample_count;
+            }
+
+            samples_in_frame.extend(input_samples.samples[start_range..end_range].iter());
+        }
+
+        // Fill at the end only if last batch is ending to quickly
+        if let Some(pts) = last_batch_end_pts {
+            if pts < end_pts + max_error {
+                Self::ensure_correct_amount_of_samples(
+                    start_pts,
+                    end_pts,
+                    sample_rate,
+                    &mut samples_in_frame,
+                );
+            }
+        }
+
+        Self::check_input_batch(start_pts, end_pts, sample_rate, &samples_in_frame);
+
+        // This call ensures that input buffer has correct amount of samples,
+        // but if it needs to do anything it is considered a bug.
+        Self::ensure_correct_amount_of_samples(
+            start_pts,
+            end_pts,
+            sample_rate,
+            &mut samples_in_frame,
+        );
 
         samples_in_frame
     }
 
-    fn samples_in_frame(start: Duration, end: Duration, sample_rate: u32) -> usize {
-        (end.saturating_sub(start).as_nanos() * sample_rate as u128 / 1_000_000_000) as usize
+    fn check_input_batch(
+        start_pts: Duration,
+        end_pts: Duration,
+        sample_rate: u32,
+        samples: &[(i16, i16)],
+    ) {
+        let samples_count_times_1e9 =
+            end_pts.saturating_sub(start_pts).as_nanos() * sample_rate as u128;
+        if samples_count_times_1e9 % 1_000_000_000 != 0 {
+            warn!(
+                "Duration {:?} is not divisible by sample duration (sample rate: {}).",
+                end_pts.saturating_sub(start_pts),
+                sample_rate,
+            )
+        }
```

Comment:
This function is called in one place - the end of the `frame_input_samples` function. I don't think this is the correct place for such validation.